src/linker/Linker/VirtualMethodAnnotationValidator.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Collections.Generic;
+using System.Diagnostics.CodeAnalysis;
+using Mono.Cecil;
+
+namespace Mono.Linker
+{
+	/// <summary>
+	/// This class validates consistency of annotations for virtual methods.
+	/// Typically the base method and its override must have the exact same annotations.
+	/// This also applies to interface method and its implementation.
+	/// </summary>
+	public class VirtualMethodAnnotationValidator
+	{
+		readonly LinkContext _context;
+		readonly HashSet<MethodDefinition> _virtualMethodsWithAnnotationsToValidate;
+
+		public VirtualMethodAnnotationValidator (LinkContext context)
+		{
+			_context = context;
+			_virtualMethodsWithAnnotationsToValidate = new HashSet<MethodDefinition> ();
+		}
+
+		public void EnqueueVirtualMethod (MethodDefinition method)
+		{
+			if (!method.IsVirtual)
+				return;
+
+			if (_context.Annotations.FlowAnnotations.RequiresDataFlowAnalysis (method) ||
+				_context.Annotations.HasLinkerAttribute<RequiresUnreferencedCodeAttribute> (method))
+				_virtualMethodsWithAnnotationsToValidate.Add (method);
+		}
+
+		public void Validate ()
+		{
+			foreach (var method in _virtualMethodsWithAnnotationsToValidate) {
+				var baseMethods = _context.Annotations.GetBaseMethods (method);
```

Comment:
How is this going to work with if the method implements 2 different interfaces which message will win?

Reply:
It will compare the methods against both of the interface methods. If both don't match it will produce two warnings. I'll add a test for that.